src/osp/Universe.cpp

```diff
@@ -49,27 +53,61 @@ void Universe::sat_remove(Satellite sat)
     m_registry.destroy(sat);
 }
 
-Vector3s Universe::sat_calc_pos(Satellite referenceFrame, Satellite target) const
+Vector3g Universe::sat_calc_pos(Satellite referenceFrame, Satellite target) const
 {
-    auto const view = m_registry.view<const UCompTransformTraj>();
+    auto const viewInCoord = m_registry.view<const UCompInCoordspace>();
+    auto const viewCoordIndex = m_registry.view<const UCompCoordspaceIndex>();
 
     // TODO: maybe do some checks to make sure they have the components
-    auto const &framePosTraj = view.get<const UCompTransformTraj>(referenceFrame);
-    auto const &targetPosTraj = view.get<const UCompTransformTraj>(target);
+    auto const &frameInCoord
+            = viewInCoord.get<const UCompInCoordspace>(referenceFrame);
+    auto const &targetInCoord
+            = viewInCoord.get<const UCompInCoordspace>(target);
 
-    if (framePosTraj.m_parent == targetPosTraj.m_parent)
+    auto const &frameCoordIndex
+            = viewCoordIndex.get<const UCompCoordspaceIndex>(referenceFrame);
+    auto const &targetCoordIndex
+            = viewCoordIndex.get<const UCompCoordspaceIndex>(target);
+
+    if (frameInCoord.m_coordSpace == targetInCoord.m_coordSpace)
     {
-        // Same parent, easy calculation
-        return targetPosTraj.m_position - framePosTraj.m_position;
+        // Both Satellites are in the same coordinate system
+
+        std::optional<CoordinateSpace> const &rSpace
+                = m_coordSpaces.at(frameInCoord.m_coordSpace);
+
+        ViewCComp_t<CCompX> const& viewX = rSpace->ccomp_view<CCompX>();
+        ViewCComp_t<CCompY> const& viewY = rSpace->ccomp_view<CCompY>();
+        ViewCComp_t<CCompZ> const& viewZ = rSpace->ccomp_view<CCompZ>();
+
+        Vector3g const pos{
+            viewX[targetCoordIndex.m_myIndex] - viewX[frameCoordIndex.m_myIndex],
+            viewY[targetCoordIndex.m_myIndex] - viewY[frameCoordIndex.m_myIndex],
+            viewZ[targetCoordIndex.m_myIndex] - viewZ[frameCoordIndex.m_myIndex],
+        };
+
+        return pos;
     }
-    // TODO: calculation for different parents
+
+    // TODO: calculation for positions across different coordinate spaces
 
     return {0, 0, 0};
 }
 
 Vector3 Universe::sat_calc_pos_meters(Satellite referenceFrame, Satellite target) const
 {
-    // 1024 units = 1 meter. this can change
-    return Vector3(sat_calc_pos(referenceFrame, target)) / 1024.0f;
+    // 1024 units = 1 meter. TODO: will change soon
+    return Vector3(sat_calc_pos(referenceFrame, target)) / gc_units_per_meter;
```

Comment:
Is the call to Vector3 constructor needed here?

If so, can this be replaced with {} ?

Reply:
This is a cast. sat_calc_pos returns 64-bit signed int Vector3g, which is casted to float and divided.

Reply:
Ah, i see. I misread the parenthesis.